docs/python/tutorial-fastapi.md

```diff
@@ -0,0 +1,652 @@
+        if item_id not in grocery_list:
+            raise HTTPException(status_code=404, detail="Item not found.")
+        # if quantity to be removed is higher or equal to item's quantity, delete the item
+        if grocery_list[item_id].quantity <= quantity:
+            del grocery_list[item_id]
+            return {"result": "Item deleted."}
+        else:
+            grocery_list[item_id].quantity -= quantity
+        return {"result": f"{quantity} items removed."}
+
+    ```
+
+12. Save the file (`kb(workbench.action.files.save)`). The application should automatically reload.
+
+You can now open the `/docs` page again and test the new routes, leveraging the debugger and the debug console to better understand the code execution. Once you're done, you can stop the debugger (`kb(workbench.action.debug.stop)`). You may also remove the breakpoint we added on step 4 by clicking on it.
+
+Congratulations! You now have a working FastAPI application with routes to add, list, and delete items from a grocery list.
+
+## Set up the data storage
+At this point, you already have a working version of the application with the base functionality. This section will guide you through setting up data storage for persistence, but you can choose to skip it if you're happy with what you've learned already.
+
+So far we are storing the data in a dictionary, which is not ideal because all of the data will be lost when the application is restarted.
+
+To persist the data, we will use [Redis](https://redis.io/), which is an open source in-memory data structure store. Due to its speed and versatility, Redis is commonly used as a data storage system in a wide range of applications, including web applications, real-time analytics systems, caching layers, this tutorial, and more.
+
+If you are already working on **GitHub Codespaces** with our existing template, you can skip directly to the ["Replace the database"](#replace-the-database) section.
+
+If you are on Windows, you can work with Redis by setting up either a [Docker container](https://www.docker.com/products/docker-desktop/), a [GitHub Codespace](https://github.com/features/codespaces) or through [WSL](https://learn.microsoft.com/en-us/windows/wsl/) (Windows Subsystem for Linux). In this tutorial we will use a Docker container, but you can refer to the [section above](#github-codespaces) for instructions on how to set up a GitHub Codespace.
+
+Otherwise, if you are on a Linux or a macOS machine, you can install Redis by following the [instructions on their website](https://redis.io/docs/getting-started/), and then skip to the ["Replace the database"](#replace-the-database) section.
+
+### Setting up a Docker Container on Windows
+The Dev Container extension for VS Code offers a simple way to configure a Linux-based workspace in your Windows machine, putting your entire project and its dependencies as well as tools you need for your project into a tidy container.
```

Comment:
I love you're bringing in the Dev Containers extension! Though I'm not sure I would introduce it as a way to set up a Linux-based environment. Would recommend pulling more from the intro in https://code.visualstudio.com/docs/devcontainers/containers and focusing on the full-featured environment part

Reply:
Could also mention the WSL extension, if you really want to focus on the Linux-based part at some point

Reply:
Great point! I tweaked it to frame it more around the "full featured dev environment" instead of "Linux-based". I mention WSL in the previous section too 😊